canvas/image.go

```diff
@@ -261,88 +263,102 @@ func (i *Image) updateReader() error {
 	return nil
 }
 
-func (i *Image) calculateMinSize() error {
-	var size fyne.Size
+func (i *Image) updateAspectAndMinSize() error {
+	var pixWidth, pixHeight int
 
 	if i.reader != nil {
-		var err error
-
-		size, err = i.minSizeFromReader(i.reader)
+		width, height, aspect, err := i.imageDetailsFromReader(i.reader)
 		if err != nil {
 			return err
 		}
+		i.aspect = aspect
+		pixWidth, pixHeight = width, height
 	} else if i.Image != nil {
 		original := i.Image.Bounds().Size()
-		size = fyne.NewSize(float32(original.X), float32(original.Y))
-		i.aspect = size.Width / size.Height
+		i.aspect = float32(original.X) / float32(original.Y)
+		pixWidth, pixHeight = original.X, original.Y
 	} else {
 		return errors.New("no matching image source")
 	}
 
 	if i.FillMode == ImageFillOriginal {
-		i.SetMinSize(size)
+		return i.setMinSizeForPixels(pixWidth, pixHeight)
 	}
 	return nil
 }
 
-func (img *Image) minSizeFromReader(source io.Reader) (fyne.Size, error) {
-	var width int
-	var height int
-
+func (i *Image) imageDetailsFromReader(source io.Reader) (width, height int, aspect float32, err error) {
 	if source == nil {
-		return fyne.NewSize(0, 0), errors.New("no matching reading reader")
+		return 0, 0, 0, errors.New("no matching reading reader")
 	}
 
-	if img.isSVG {
+	if i.isSVG {
 		var err error
 
-		img.icon, err = svg.NewDecoder(source)
+		i.icon, err = svg.NewDecoder(source)
 		if err != nil {
-			return fyne.NewSize(0, 0), err
+			return 0, 0, 0, err
 		}
-		config := img.icon.Config()
+		config := i.icon.Config()
 		width, height = config.Width, config.Height
-		img.aspect = config.Aspect
+		aspect = config.Aspect
 	} else {
 		var buf bytes.Buffer
 		tee := io.TeeReader(source, &buf)
-		img.reader = io.MultiReader(&buf, source)
+		i.reader = io.MultiReader(&buf, source)
 
 		config, _, err := image.DecodeConfig(tee)
 		if err != nil {
-			return fyne.NewSize(0, 0), err
+			return 0, 0, 0, err
 		}
 		width, height = config.Width, config.Height
-		img.aspect = float32(width) / float32(height)
+		aspect = float32(width) / float32(height)
 	}
+	return
+}
 
-	dpSize, err := scale.ToFyneSize(img, width, height)
+func (i *Image) setMinSizeForPixels(width, height int) error {
+	app := fyne.CurrentApp()
+	if app == nil {
+		return nil // error logged already with more info
+	}
+	driver := app.Driver()
+	if driver == nil {
+		return errors.New("no current driver")
+	}
+	c := driver.CanvasForObject(i)
+	if c == nil {
+		return nil // this will happen a lot during init
+	}
+	dpSize, err := scale.ToFyneSize(i, width, height)
```

Comment:
scale.ToFyneSize does all the access check above (fyne.CurrentApp().Driver().CanvasForObject(i) isn't necessary anymore)